Feat/http delete

vdedios opened the pull request:
Solver #50 

Behaviour has been taken from [ Nginx http dav module](https://nginx.org/en/docs/http/ngx_http_dav_module.html) which is the one implementing methods related with modifiying resources.

vdedios commented:
I know some tests are breaking :cry:  , will investigate it tomorrow

Damrod commented:
I've been running manually a couple of tests to see how your feature works. I'm not sure I understand what is it that the integration test that you wrote is supposed to do, but
1. as far as I can see I have to actually create (mkdir) the upload folder in order to actually create the file with the http://localhost:8084/delete_file.html "create file named" button inside of it.
2. If I chmod 000, and chown root:root to no_privileges_folder/test.txt file, and then click on delete in the same page, it still deletes the file. This is disconcerting. I am running the program webserv as my unprivileged user. I wouldn't have expected the deletion to succeed.

vdedios commented:
@Damrod there was an error on my side. I accidentally deleted the `upload` folder and add it to the `.gitignore`. This is why you should create it first but is no longer needed since I added an exception the `.gitignore` file. This also fixed the `delete` tests. The html file I've created is just to upload a simple test file to test deletion after.

The deletion with a non-privileged user issue also drove me crazy. For what I've been investigating, you need to log with an unprivileged user which does not belong either to the same group of the file or the `sudo` group (which you probably belong to, this was also happening to me. You can check it whith `groups` command) Meeting this two conditions you shouldn't be able to delete the file.

Damrod commented on srcs/app/http/response/delete/HttpDeleteResponse.cpp:
```diff
@@ -3,18 +3,64 @@
+void	HttpDeleteResponse::Delete_(File const &file) {
+	if (file.IsDirectory()) {
+		BuildResponse_(
+			nftw(file.GetPath().c_str(),
```
So the idea behind this ntfw is to remove all contents in a directory?

vdedios replied:
Yes, I also though about it. Since `nftw` is a function which main purpose is to traverse the directory tree and applying a function (in this case `RemoveSingleFile_`) for me was more legible to distinguish between the two cases: one having to walk through al the file hierarchy and then a case for deleting simple file. But yes, it should also work :)

Damrod commented:
> @Damrod there was an error on my side. I accidentally deleted the `upload` folder and add it to the `.gitignore`. This is why you ahd to create it first. This is no longer needed since I added an exception the `.gitignore` file. This also fixed the `delete` tests. The html file I've created is just to upload a simple test file to test deletion after.
> 
> The deletion with a non-privileged user issue also drove me crazy. For what I've been investigating, you need to log with an unprivileged user which does not belong either to the same group of the file or the `sudo` group (which you probably belong to, this was also happening to me. You can check the groups you belong to whith `groups` command) Meeting this two conditions you shouldn't be able to delete the file. I mean, this should also happen if you try to do a simple `rm file_name` with that user you are triying with.

Thanks for the explanation! It works! Great job, congrats

Damrod reviewed (approved):
Great job! Thank you!

gbudau commented on srcs/app/http/response/delete/HttpDeleteResponse.cpp:
```diff
@@ -3,18 +3,64 @@
+	if (!typeflag) (void)typeflag;
+	if (!ftwbuf) (void)ftwbuf;
+
+    return remove(fpath);
```
```suggestion
    remove(fpath);
    return 0;
```

I believe this function should always `return 0`, and we could check `errno` for errors.


Currently, if `nftw()` fails while removing a non-empty directory, it leaves it in an inconsistent state.

After the function returns, the directory will contain files that could not delete and files that it could have deleted, but it didn't delete because it exited early.

The module [ngx_http_dav_module](https://nginx.org/en/docs/http/ngx_http_dav_module.html) returns a `500` response in this cases, and deletes all the files/directories that can be deleted.

vdedios replied:
Seems a good idea. The only problem here would be that `errno` value could get overwritten. What happen if many files fail? Lets imagine one fails with a permission error and another one with another kind, what should you respond?

As said some comments above I think we should take `Nginx` as a reference but with some flexibility, making a consistent implementation.

That being said, deleting all possible files seem ok for me but we should define how to handle a case where many files fail for different circunstances, what code should we return?

gbudau replied:
Since this function is recursive and the last directory is the parameter, I think the last `errno` value will be `EEXIST` or `ENOTEMPTY`.

`500` seems like a decent response status in this case.

vdedios replied:
I've been reviewing this and the problem with that approach is you can't know if an error has ocurred because, if you let `nftw` keep running, you'll end up (if nothing else happens) with a success return status. This way, you won't have a reason to check `errno` value. A work around this could be keeping track of every return status of `RemoveSingleFile_` function storing the result on a private variable or something similar. But, since this is a static function, you'll end up sharing states with different connection calls...Probably there is a way to fix this but, since it works and its a matter of just imitating at most `Nginx` behaviour, for me is good enough and I'll leave it as is for now :sweat_smile: Hope I've just convinced you a little bit :smile:

gbudau commented on srcs/app/test/integration/delete_method_test.py:
```diff
@@ -0,0 +1,103 @@
+import os
```
Great idea to split the tests into multiple modules. Thank you!

gbudau commented on srcs/app/test/integration/delete_method_test.py:
```diff
@@ -0,0 +1,103 @@
+import tempfile
+import time
+
+PROJ_DIR = str(Path(__file__).parents[4])
+TMP_UPLOAD_DIR = PROJ_DIR + '/html/web3/upload/'
+
+@pytest.fixture(scope='module', autouse=True)
+def start_webserv():
+    webserv = subprocess.Popen('./webserv', cwd=PROJ_DIR)
+    time.sleep(2)
+    yield webserv
+    webserv.terminate()
+
```
The [pytest documentation](https://docs.pytest.org/en/6.2.x/fixture.html#scope-sharing-fixtures-across-classes-modules-packages-or-session) recommends creating a `conftest.py` file and putting all the fixtures there so that tests from multiple test modules in the directory can access the fixture functions.

This way, we could also add `scope="session"` to some of the fixtures to only be invoked once per test session, and we could import the variables from other modules.
```
from conftest import PROJ_DIR
```
<br>

A useful command to see the setup/teardown of the tests:
```
pytest --setup-show
```

vdedios replied:
Sounds good, though I see this an improvement for another task. If you want to go with it ;) Just a short reminder for this, there is a "problem" with delete tests. Since those are deleting files, you cannot re-use those fixtures and have to create new ones.

gbudau replied:
Ok, I will add a PR for this task after we merge this.

gbudau commented on srcs/app/test/integration/delete_method_test.py:
```diff
@@ -0,0 +1,103 @@
+
+# ERROR RESPONSES
+
+@pytest.fixture(scope='module')
+def tmp_dir_no_delete():
+    yield tempfile.mkdtemp(dir = TMP_UPLOAD_DIR)
+
```
This fixture looks like a copy of `tmp_dir` and `tmp_nested_dir`. Why do we need them?

vdedios replied:
As said, this fixtures get deleted, so you need to create new ones. This is the first time I use this library so for sure there'll be better ways of doing it

gbudau reviewed (approved):
Great job! Overall it looks good. Thank you!

vdedios commented on srcs/app/http/response/delete/HttpDeleteResponse.cpp:
```diff
@@ -3,18 +3,64 @@
+}
+
+int	HttpDeleteResponse::RemoveSingleFile_(
+								const char *fpath,
+								const struct stat *sb,
+								int typeflag,
+								struct FTW *ftwbuf) {
+	// cast to void when function not used as callback
+	if (!sb) (void)sb;
+	if (!typeflag) (void)typeflag;
+	if (!ftwbuf) (void)ftwbuf;
```
I don't know, I have mixed feelings about it because you should keep `fpath` which is being used :thinking:. So, having some params named and others don't seems a little bit strange. But I see your point

gbudau replied:
I forgot to keep the `fpath` name :sweat_smile: . For me, it compiles and works correctly. I get your point, and I'm ok with whatever you decide.

Damrod replied:
you could also do it with __attribute__((unused)):
```
diff --git a/srcs/app/http/response/delete/HttpDeleteResponse.cpp b/srcs/app/http/response/delete/HttpDeleteResponse.cpp
index 2443be4..0fcdcd6 100644
--- a/srcs/app/http/response/delete/HttpDeleteResponse.cpp
+++ b/srcs/app/http/response/delete/HttpDeleteResponse.cpp
@@ -36,16 +36,11 @@ void	HttpDeleteResponse::Delete_(File const &file) {
 }
 
 int	HttpDeleteResponse::RemoveSingleFile_(
-								const char *fpath,
-								const struct stat *sb,
-								int typeflag,
-								struct FTW *ftwbuf) {
-	// cast to void when function not used as callback
-	if (!sb) (void)sb;
-	if (!typeflag) (void)typeflag;
-	if (!ftwbuf) (void)ftwbuf;
-
-    return remove(fpath);
+						const char *fpath,
+						__attribute__((unused)) const struct stat *sb,
+						__attribute__((unused)) int typeflag,
+						__attribute__((unused)) struct FTW *ftwbuf) {
+	return remove(fpath);
 }
 
 void	HttpDeleteResponse::BuildResponse_(int return_status) {
```

vdedios replied:
Adding @Damrod suggestion since it is a little bit more explicit and self-explanatory. Both compiled ok!

vdedios commented on srcs/app/http/response/delete/HttpDeleteResponse.cpp:
```diff
@@ -3,18 +3,64 @@
+			? SetErrorRawResponse_(403)
+			: SetErrorRawResponse_(500);
+	} else {
+		DefaultStatusResponse_(200);
```
Yes, I know about it. `204` is a succesful response without body. Since we own this webserver and we are only meant to have `Nginx` as an inspiration I found useful to return a small templated response, may be useful from some scenarios. I am also sure that `Nginx` allows to configure some kind of templated response but, from my point of view, this was a little out of the scope of this feature :cold_sweat:

vdedios commented on config/default.conf:
```diff
@@ -38,13 +38,13 @@ server {
 	location	/upload/ {
 		client_max_body_size 100000000;
-		limit_except POST GET;
+		limit_except POST GET DELETE;
```
Ok, will add it back again! I was just tinkering around with the config files :nerd_face:

vdedios commented on html/web3/delete_file.html:
```diff
@@ -0,0 +1,66 @@
+        })
+    }
+
+    function switchStates() {
+        delButton.disabled ^= true;
+        genButton.disabled ^= true;
+        delFile.disabled ^= true;
+        genFile.disabled ^= true;
+    }
```
Yes, it may be not very readable and I am sure there are many more elegant ways of doing it. Just wanted to build a quick example. Sorry for the inconveniences :sweat_smile:

gbudau replied:
I was not referring to the xor :sweat_smile:. I found it annoying because it allows only create/delete/create/etc, and I wanted to test create/create/create/delete/delete/etc. Because of this, I deleted it while trying out the feature.

Damrod replied:
I have to say that I agree with gbudau. Why create/delete/create... instead of create/create/create/delete/delete???

I hoped the +1 in the previous post would have conveyed this...

vdedios replied:
Oups, missed this one. Will ammend it in a new PR

vdedios replied:
There you go @Damrod https://github.com/42AGV/webserv/pull/136

vdedios commented on .github/workflows/integration_test_check.yml:
```diff
@@ -28,6 +28,14 @@ jobs:
+          if [ "$RUNNER_OS" == "Linux" ]; then \
+          sudo su -c "useradd test_user -s /bin/bash -m"
+          sudo chown -R test_user:test_user no_privileges_folder
+          fi
```
This is much better thanks! I did not knew about this "mock" user :laughing:

vdedios commented:
Closing and merging PR, thanks both of you for your comments and time!!
